src/audio/base_fw.c

```diff
@@ -323,6 +325,68 @@ static uint32_t basefw_get_ext_system_time(uint32_t *data_offset, char *data)
 	return IPC4_UNAVAILABLE;
 }
 
+static int basefw_register_kcps(bool first_block,
+				bool last_block,
+				uint32_t data_offset_or_size,
+				const char *data)
+{
+	if (!(first_block && last_block))
+		return IPC4_INVALID_REQUEST;
+
+	if (core_kcps_adjust(0, *(int *)data))
+		return IPC4_FAILURE;
+
+	return IPC4_SUCCESS;
+}
+
+static int basefw_KCPS_allocation_request(struct ipc4_resource_kcps *request)
+{
+	if (core_kcps_adjust(request->core_id, request->kcps))
+		return IPC4_FAILURE;
+
+	return IPC4_SUCCESS;
+}
+
+static int basefw_resource_allocation_request(bool first_block,
+					      bool last_block,
+					      uint32_t data_offset_or_size,
+					      const char *data)
+{
+	struct ipc4_resource_request *request;
+
+	if (!(first_block && last_block))
+		return IPC4_INVALID_REQUEST;
+
+	request = (struct ipc4_resource_request *)data;
+
+	switch (request->ra_type) {
+	case IPC4_RAT_DSP_KCPS:
+		return basefw_KCPS_allocation_request(&request->ra_data.kcps);
+	case IPC4_RAT_MEMORY:
+		return IPC4_UNKNOWN_MESSAGE_TYPE;
+	default:
+		return IPC4_UNKNOWN_MESSAGE_TYPE;
+	}
+}
+
+static int basefw_power_state_info_get(uint32_t *data_offset, char *data)
+{
+	struct ipc4_power_state_info *power_state_info = (struct ipc4_power_state_info *)data;
+
+	power_state_info->type1 = ACTIVE_CORES_MASK;
+	power_state_info->len1 = sizeof(power_state_info->active_cores_mask);
+	power_state_info->active_cores_mask = cpu_enabled_cores();
+	power_state_info->type2 = CORE_KCPS;
+	power_state_info->len2 = CONFIG_CORE_COUNT * sizeof(uint32_t);
+
+	for (int i = 0; i < CONFIG_CORE_COUNT; i++)
+		power_state_info->core_kcps[i] = core_kcps_get(i);
+
+	*data_offset = sizeof(struct ipc4_power_state_info);
+
+	return IPC4_SUCCESS;
```

Comment:
**struct _ipc4_power_state_info_** contains Type + Length + Value, this looks like TVL, so maybe use tlv.

```c
struct ipc4_tuple *tuple = (struct ipc4_tuple *)data;
uint32_t core_kcps[CONFIG_CORE_COUNT] = {0, 0, 0};

set_tuple_uint32(tuple, ACTIVE_CORES_MASK, cpu_enabled_cores());
tuple = next_tuple(tuple);

for (int core_id = 0; core_id < CONFIG_CORE_COUNT; core_id++) {
    if (arch_cpu_active(core_id))
        core_kcps[core_id] = core_kcps_get(core_id);
}

set_tuple(tuple, CORE_KCPS, sizeof(core_kcps), core_kcps);
tuple = next_tuple(tuple);
*data_offset = (int)((char *)tuple - data);
return IPC4_SUCCESS;
```

Reply:
thx, didn't notice tuple there being TLV.